core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -707,8 +707,11 @@ public Result visit(Values e) {
       } else if (list.size() == 1) {
         query = list.get(0);
       } else {
-        query = SqlStdOperatorTable.UNION_ALL.createCall(
-            new SqlNodeList(list, POS));
+        SqlNode sqlNode = SqlStdOperatorTable.UNION_ALL.createCall(POS, list.get(0), list.get(1));
+        for (int i = 2; i < list.size(); i++) {
+          sqlNode = SqlStdOperatorTable.UNION_ALL.createCall(POS, sqlNode, list.get(i));
```

Comment:
What this does is instead of creating a SINGLE UNION OPERATOR with a FLAT LIST of all SUB-SELECTS, it creates it iteratively like a BINARY OPERATOR as it should be as UNION_ALL operator is a BINARY Operator.

Reply:
Yes, previously it create SINGLE UNION operator on list of operands which eventually fail at unparse logic as UNION ALL is binary operator and does not support multiple operands as inputs.